### DISCUSSION TITLE
Improve the E-Commerce Template

### DISCUSSION THREAD

**Discussion by jacobsfletch** [payloadcms/payload#4352] [Feature Requests & Ideas]
Our [E-commerce Template](https://github.com/payloadcms/payload/tree/main/templates/ecommerce) has been wildly successful and is currently being used as a starting point for a great number of new projects. But while this template is designed to be minimal yet comprehensive, giving you a boilerplate to build something completely unique—it _does_ lack in a few notable areas. So we need to take some time to build all the rest of the features that were left out of the initial build.

This discussion will be a place to log all of the work that needs to be done in the short-term. It can be treated as a living document where we continue to adjust and refine the scope over time based on direct feedback we get from the community. If you have suggestions that you think are valuable enough to build into this template, or someone else has already suggested your feature, please leave a comment of support and we'll consider adding it into the list.

One last thing to quickly note here is that there is a real effort to make these templates as simple as possible. Templates are meant to demonstrate Payload in a specific capacity, and to build _on top of_ without tearing down existing code. That being said, they need to remain as agnostic as possible, in both design and functionality (without sacrificing quality or lacking features).

Here's a simple list to get the ball rolling, I'll personally add more as they pop into my mind:

Features:
- Dynamic product attributes: ability to set arbitrary product details that could potentially effect price or product image, such as color or material.
- Shipping details: we currently are not collecting shipping details during checkout, but in reality, are a requirement of most online stores who sell physical goods.
- Inventory tracking: ability to prevent sold-out items from continuing to sell, and indicate to a shopper that an item is low in stock.
- Alternative payment processing: provide alternatives to Stripe along with clear instruction on how to swap in your own payment processor such as PayPal.
- URL in state: push state out of local React components and into the URL as much as possible, i.e. search params for the `ArchiveBlock`.

Improvements:
- Tailwind CSS?: I know how much everyone loves Tailwind. Should we add it?
- Design polish: without getting _too_ design-y here, I think a mere 10% polish to design would go a _long_ way. 

More to come. Please join the conversation and/or upvote this discussion if you think this work should be prioritized!

**Comment by HarleySalas** [comment]
I believe the goal should be to eventually move away from Stripe. Not all countries can use it and even then, most enterprise businesses, do **not** want to use Stripe. Even over that, it is forcing vendor lock-in. If there is a multi-million dollar, or even billion dollar business that wants to build an e-commerce site, they will not lock-in to a payment processor.

Every element should be completely agnostic of any specific corporation. Allow us to integrate any payment processor, any shipping provider, etc.

While I _wish_ you guys would focus on Payload itself, rather than an e-commerce template. (I would love to see a Medusa plugin!), you could still look at Medusa for inspiration, or even use the Medusa modules themselves as a framework to build a "**payload-plugin-ecommerce**," which stands on it's own.

E-commerce is an extremely complex thing and IF you figure out a way to integrate Medusa, you'll benefit from the millions of dollars and full-time effort that they have going on towards enterprise e-commerce. There is a **LOT** of value that is coming from them. Just like you're capitalizing on Drizzle now for databases, you could be capitalizing on Medusa for what they're doing in the e-commerce world. It's **really** worth evaluating. No one is providing first-class support for Medusa. You guys have the power to make a Medusa site as easy as a Shopify site.

**Reply by jacobsfletch** [reply]
Thank you @HarleySalas 🙏

> I believe the goal should be to eventually move away from Stripe. 

I agree we should find better pattern to make it _easier_ to swap out payment processors, especially in the spirit of open-source and headless. But we won't want to _remove_ our Stripe integration. We just need to provide alternatives here, along with clear instruction.

> I would love to see a Medusa plugin

This would make a killer plugin. If you're reading this and you're well-versed in Medusa and Payload, maybe _you_ could lead the effort! Community-supported plugins have always been well-received and can be loads of fun, here's how: https://payloadcms.com/blog/tutorial-building-your-own-payload-plugin

>  It's really worth evaluating.

This is all great and I completely agree. I'm going to add this to the list above 👍

**Comment by HarleySalas** [comment]
> > I would love to see a Medusa plugin
> 
> This would make a killer plugin. If you're reading this and you're well-versed in Medusa and Payload, maybe _you_ could lead the effort! Community-supported plugins have always been well-received and can be loads of fun, here's how: https://payloadcms.com/blog/tutorial-building-your-own-payload-plugin
> 

I'm not opposed to giving it a try, but I'm not sure I'm the right person to develop it. I've worked in e-commerce for a long time. While I am a developer now, I'm by no means a great developer. I work on the business side and I own an agency.

Perhaps there should be a discussion as to exactly _what_ a Medusa plugin would look like long before we start actually making one. I tried looking at the Medusa plugin for Strapi, but couldn't even get it running. It doesn't seem to be very well maintained. 

As far as I understand, they are essentially importing everything from Medusa in to Strapi, then any time something is created, or updated on either side, they sync.

To be honest, that idea terrifies me. Maybe it shouldn't... But, I also really like a single source of truth and not having the potential for some sort of mess.

I'm not sure if Payload should be looking to actually be the fulfillment dashboard, as Medusa does a very good job of this and it's constantly being improved. Perhaps Payload should just be doing the content management side of things, rather than looking to be a replacement for the medusa dashboard.

I hope you guys get the time to checkout medusa; install it's starter, take a look at it's features and tell me what you guys think would be the best way forward in that regard.

**Comment by HarleySalas** [comment]
After some conversations in the Discord, I think I'm pretty convinced that;

Medusa does E-Commerce very well. It is their primary focus.

Payload does content management very well, even though it is no longer it's primary focus. (I tweeted about syntax+payload and I _think_ that tweet resulted in the episode? @jmikrut 😂 I heard you about getting away from the notion; great podcast.) And while Payload isn't just a CMS anymore, I don't think you guys are looking to invest heavily in E-Commerce?

So while it could be very possible to use Medusa's modules as a framework, particularly with [Medusa v2](https://medusajs.com/blog/toward-medusa-2/) coming out; to build a "payload-plugin-ecommerce", using Medusa under the hood, it's probably quite unnecessary.

Sure, it'd be nice to use one dashboard and only deploy one thing, but we lose a lot of value, or assume a lot of responsibility if we try to really combine them entirely. 

So, this leads me to the idea of just having a `payload-plugin-medusa`, but that comes with it's own concerns. Currently, there are only 2 CMS plugins that I know of; Contentful and Strapi. As far as I know, they work by importing your medusa products and syncing both sides when either side has an update. I **hate** this idea and it seems like it could lead to a lot of problems.

So, would it be possible to combine Payload's local api with Medusa's [admin api](https://docs.medusajs.com/api/admin) in such a way that there was only a single source of truth for each piece of data involved? Payload + Payload Admin could focus on what it does best (managing content and extending) and Medusa could focus on what it does best (E-commerce business logic and fulfillment dashboard)

It sounds a bit like an odd approach perhaps, but I think it makes the most sense for e-commerce. If you guys could pull off making medusa content management simple and somehow make both products work well together, it could gain a **lot** of traction for Payload and I would certainly be pitching it **everywhere**, proudly.

**Comment by r1tsuu** [comment]
Honestly i don't know, i think templates should be as simple as possible.
I'm building E-Commerce site now with PayloadCMS and i just can't use some full stack template because i have specific design and project requirements.
And about Medusa, as i know it's a separate CMS for managing products/orders, why exactly would you want to combine that with Payload Dashboard?
If it's a case for example when you need to show some products from Medusa in your blog post that managed by Payload, you can easily do that with custom field component.
But would some imaginary plugin work for this very specific case? You can't just put into the plugin everything, actually it's the same problem as with site template with many stuff by default. That's why you need to know how to do it by yourself first.. Maybe here we should think about some 'Headless" plugin that just gives you react hooks and functions for endpoints/server hooks to CRUD data from Medusa.
Of course there are cases when you just want quickly build a E-commerce with some basic stuff, but i don't think Payload as code first CMS designed for this.

**Reply by jacobsfletch** [reply]
> And about Medusa, as i know it's a separate CMS for managing products/orders, why exactly would you want to combine that with Payload Dashboard?

These are my thoughts exactly, without knowing too much about Medusa, from what I saw it _does_ seem quite identical to the E-commerce Template. It's just a dashboard with a payment processor, minus Payload ftl. They even offer a Stripe integration (along with PayPal, which is exactly what I think we should do).

**Comment by acandael** [comment]
Hi Jacob,

I'm really thrilled about the Ecommerce template for Payload CMS. The only bummer is that I don't see any functionality for product variants. So I wonder if this feature is planned for a future version of the Ecommerce template. Great work so far!

**Reply by jacobsfletch** [reply]
> I'm really thrilled about the Ecommerce template for Payload CMS. The only bummer is that I don't see any functionality for product variants. So I wonder if this feature is planned for a future version of the Ecommerce template. Great work so far!

I fully agree and I know others would as well. I've added products variants to the list in the description 👍 that one is pretty essential.

**Comment by Allanio** [comment]
Hi @jacobsfletch,

Some basic level of integration with Printify APIs for print-on-demand would be huge for those who do not want to manage inventory. I think this will unlock instant value for many out the box for this E-commerce template and perfectly demonstrate what can be achieved end-to-end.

**Comment by paulpopus** [comment]
Closing this discussion in light of ecommerce plugin and template being released. We're going to capture discussions going forward in this RFC so we can better manage engagement on this topic: https://github.com/payloadcms/payload/discussions/14239

Thank you everyone!
